File: backend/app/services/shipping_rates.py
from __future__ import annotations

from dataclasses import dataclass
from math import ceil

from app.schemas.cost import OriginRegion


@dataclass(frozen=True)
class ShippingRateQuote:
    origin_region: OriginRegion
    rate_band: str
    rate_weight_kg: float
    shipping_fee: float


WEST_MALAYSIA_DOCUMENT_RATES = {
    0.5: 64.90,
    1.0: 79.20,
    1.5: 93.30,
    2.0: 116.38,
}

WEST_MALAYSIA_PARCEL_RATES = {
    0.5: 64.90,
    1.0: 79.20,
    1.5: 93.30,
    2.0: 116.38,
    2.5: 139.65,
    3.0: 167.58,
    3.5: 201.10,
    4.0: 216.10,
    4.5: 231.10,
    5.0: 246.10,
    5.5: 261.10,
    6.0: 276.10,
    6.5: 291.10,
    7.0: 306.10,
    7.5: 321.10,
    8.0: 336.10,
    8.5: 351.10,
    9.0: 366.10,
    9.5: 381.10,
    10.0: 396.10,
    10.5: 411.10,
    11.0: 426.10,
    11.5: 441.10,
    12.0: 456.10,
    12.5: 471.10,
    13.0: 486.10,
    13.5: 501.10,
    14.0: 516.10,
    14.5: 531.10,
    15.0: 546.10,
    15.5: 561.10,
    16.0: 576.10,
    16.5: 591.10,
    17.0: 606.10,
    17.5: 621.10,
    18.0: 636.10,
    18.5: 651.10,
    19.0: 666.10,
    19.5: 681.10,
    20.0: 696.10,
    20.5: 711.10,
    21.0: 726.10,
    21.5: 741.10,
    22.0: 756.10,
    22.5: 771.10,
    23.0: 786.10,
    23.5: 801.10,
    24.0: 816.10,
    24.5: 831.10,
    25.0: 846.10,
    25.5: 861.10,
    26.0: 876.10,
    26.5: 891.10,
    27.0: 906.10,
    27.5: 921.10,
    28.0: 936.10,
    28.5: 951.10,
    29.0: 966.10,
    29.5: 981.10,
    30.0: 996.10,
}

EAST_MALAYSIA_DOCUMENT_RATES = {
    0.5: 77.87,
    1.0: 89.55,
    1.5: 107.46,
    2.0: 139.70,
}

EAST_MALAYSIA_PARCEL_RATES = {
    0.5: 77.87,
    1.0: 89.55,
    1.5: 107.46,
    2.0: 139.70,
    2.5: 167.64,
    3.0: 201.17,
    3.5: 241.40,
    4.0: 261.40,
    4.5: 281.40,
    5.0: 301.40,
    5.5: 321.40,
    6.0: 341.40,
    6.5: 361.40,
    7.0: 381.40,
    7.5: 401.40,
    8.0: 421.40,
    8.5: 441.40,
    9.0: 461.40,
    9.5: 481.40,
    10.0: 501.40,
    10.5: 521.40,
    11.0: 541.40,
    11.5: 561.40,
    12.0: 581.40,
    12.5: 601.40,
    13.0: 621.40,
    13.5: 641.40,
    14.0: 661.40,
    14.5: 681.40,
    15.0: 701.40,
    15.5: 721.40,
    16.0: 741.40,
    16.5: 761.40,
    17.0: 781.40,
    17.5: 801.40,
    18.0: 821.40,
    18.5: 841.40,
    19.0: 861.40,
    19.5: 881.40,
    20.0: 901.40,
    20.5: 921.40,
    21.0: 941.40,
    21.5: 961.40,
    22.0: 981.40,
    22.5: 1001.40,
    23.0: 1021.40,
    23.5: 1041.40,
    24.0: 1061.40,
    24.5: 1081.40,
    25.0: 1101.40,
    25.5: 1121.40,
    26.0: 1141.40,
    26.5: 1161.40,
    27.0: 1181.40,
    27.5: 1201.40,
    28.0: 1221.40,
    28.5: 1241.40,
    29.0: 1261.40,
    29.5: 1281.40,
    30.0: 1301.40,
}


def quote_shipping_rate(origin_region: OriginRegion | str, weight_kg: float) -> ShippingRateQuote:
    region = OriginRegion(origin_region)
    weight_band = _weight_band(weight_kg)
    rounded_weight = _round_up_to_half(weight_kg)
    rate_table = _rate_table(region, weight_band)
    if not rate_table:
        raise ValueError("No shipping rates configured")

    rate_weight = min(max(rounded_weight, min(rate_table)), max(rate_table))
    shipping_fee = rate_table[rate_weight]
    return ShippingRateQuote(
        origin_region=region,
        rate_band=weight_band,
        rate_weight_kg=rate_weight,
        shipping_fee=round(shipping_fee, 2),
    )


SEA_MIN_CHARGEABLE_WEIGHT_KG = 20.0
SEA_MAX_TABLE_WEIGHT_KG = 32.0
SEA_WEIGHT_STEP_KG = 0.5


SEA_WEST_MALAYSIA_LCL_RATES = {
    round(20.0 + (index * SEA_WEIGHT_STEP_KG), 1): round(150.0 + (index * 3.0), 2)
    for index in range(int((SEA_MAX_TABLE_WEIGHT_KG - SEA_MIN_CHARGEABLE_WEIGHT_KG) / SEA_WEIGHT_STEP_KG) + 1)
}

SEA_EAST_MALAYSIA_LCL_RATES = {
    round(20.0 + (index * SEA_WEIGHT_STEP_KG), 1): round(200.0 + (index * 4.5), 2)
    for index in range(int((SEA_MAX_TABLE_WEIGHT_KG - SEA_MIN_CHARGEABLE_WEIGHT_KG) / SEA_WEIGHT_STEP_KG) + 1)
}


def quote_sea_shipping_rate(origin_region: OriginRegion | str, weight_kg: float) -> ShippingRateQuote:
    region = OriginRegion(origin_region)
    rounded_weight = _round_up_to_half(weight_kg)
    billable_weight = min(max(rounded_weight, SEA_MIN_CHARGEABLE_WEIGHT_KG), SEA_MAX_TABLE_WEIGHT_KG)
    table = SEA_EAST_MALAYSIA_LCL_RATES if region == OriginRegion.east else SEA_WEST_MALAYSIA_LCL_RATES
    shipping_fee = table[billable_weight]
    return ShippingRateQuote(
        origin_region=region,
        rate_band="sea_lcl",
        rate_weight_kg=billable_weight,
        shipping_fee=round(shipping_fee, 2),
    )


def _weight_band(weight_kg: float) -> str:
    return "document" if weight_kg <= 2.0 else "parcel"


def _round_up_to_half(weight_kg: float) -> float:
    return round(ceil(weight_kg * 2.0) / 2.0, 2)


def _rate_table(region: OriginRegion, weight_band: str) -> dict[float, float]:
    if region == OriginRegion.east:
        return EAST_MALAYSIA_DOCUMENT_RATES if weight_band == "document" else EAST_MALAYSIA_PARCEL_RATES
    return WEST_MALAYSIA_DOCUMENT_RATES if weight_band == "document" else WEST_MALAYSIA_PARCEL_RATES

File: frontend/src/lib/shippingRates.ts
export type OriginRegion = "west" | "east";

export type ShippingRateBand = "document" | "parcel" | "sea_lcl";

export type ShippingRateQuote = {
  originRegion: OriginRegion;
  rateBand: ShippingRateBand;
  rateWeightKg: number;
  shippingFee: number;
};

const WEST_MALAYSIA_DOCUMENT_RATES: Record<number, number> = {
  0.5: 64.9,
  1: 79.2,
  1.5: 93.3,
  2: 116.38,
};

const WEST_MALAYSIA_PARCEL_RATES: Record<number, number> = {
  0.5: 64.9,
  1: 79.2,
  1.5: 93.3,
  2: 116.38,
  2.5: 139.65,
  3: 167.58,
  3.5: 201.1,
  4: 216.1,
  4.5: 231.1,
  5: 246.1,
  5.5: 261.1,
  6: 276.1,
  6.5: 291.1,
  7: 306.1,
  7.5: 321.1,
  8: 336.1,
  8.5: 351.1,
  9: 366.1,
  9.5: 381.1,
  10: 396.1,
  10.5: 411.1,
  11: 426.1,
  11.5: 441.1,
  12: 456.1,
  12.5: 471.1,
  13: 486.1,
  13.5: 501.1,
  14: 516.1,
  14.5: 531.1,
  15: 546.1,
  15.5: 561.1,
  16: 576.1,
  16.5: 591.1,
  17: 606.1,
  17.5: 621.1,
  18: 636.1,
  18.5: 651.1,
  19: 666.1,
  19.5: 681.1,
  20: 696.1,
  20.5: 711.1,
  21: 726.1,
  21.5: 741.1,
  22: 756.1,
  22.5: 771.1,
  23: 786.1,
  23.5: 801.1,
  24: 816.1,
  24.5: 831.1,
  25: 846.1,
  25.5: 861.1,
  26: 876.1,
  26.5: 891.1,
  27: 906.1,
  27.5: 921.1,
  28: 936.1,
  28.5: 951.1,
  29: 966.1,
  29.5: 981.1,
  30: 996.1,
};

const EAST_MALAYSIA_DOCUMENT_RATES: Record<number, number> = {
  0.5: 77.87,
  1: 89.55,
  1.5: 107.46,
  2: 139.7,
};

const EAST_MALAYSIA_PARCEL_RATES: Record<number, number> = {
  0.5: 77.87,
  1: 89.55,
  1.5: 107.46,
  2: 139.7,
  2.5: 167.64,
  3: 201.17,
  3.5: 241.4,
  4: 261.4,
  4.5: 281.4,
  5: 301.4,
  5.5: 321.4,
  6: 341.4,
  6.5: 361.4,
  7: 381.4,
  7.5: 401.4,
  8: 421.4,
  8.5: 441.4,
  9: 461.4,
  9.5: 481.4,
  10: 501.4,
  10.5: 521.4,
  11: 541.4,
  11.5: 561.4,
  12: 581.4,
  12.5: 601.4,
  13: 621.4,
  13.5: 641.4,
  14: 661.4,
  14.5: 681.4,
  15: 701.4,
  15.5: 721.4,
  16: 741.4,
  16.5: 761.4,
  17: 781.4,
  17.5: 801.4,
  18: 821.4,
  18.5: 841.4,
  19: 861.4,
  19.5: 881.4,
  20: 901.4,
  20.5: 921.4,
  21: 941.4,
  21.5: 961.4,
  22: 981.4,
  22.5: 1001.4,
  23: 1021.4,
  23.5: 1041.4,
  24: 1061.4,
  24.5: 1081.4,
  25: 1101.4,
  25.5: 1121.4,
  26: 1141.4,
  26.5: 1161.4,
  27: 1181.4,
  27.5: 1201.4,
  28: 1221.4,
  28.5: 1241.4,
  29: 1261.4,
  29.5: 1281.4,
  30: 1301.4,
};

const SEA_MIN_CHARGEABLE_WEIGHT_KG = 20;
const SEA_MAX_TABLE_WEIGHT_KG = 32;
const SEA_WEIGHT_STEP_KG = 0.5;

const SEA_WEST_MALAYSIA_LCL_RATES: Record<number, number> = Object.fromEntries(
  Array.from({ length: ((SEA_MAX_TABLE_WEIGHT_KG - SEA_MIN_CHARGEABLE_WEIGHT_KG) / SEA_WEIGHT_STEP_KG) + 1 }, (_, index) => {
    const weight = Number((SEA_MIN_CHARGEABLE_WEIGHT_KG + (index * SEA_WEIGHT_STEP_KG)).toFixed(1));
    const price = Number((150 + (index * 3)).toFixed(2));
    return [weight, price];
  })
);

const SEA_EAST_MALAYSIA_LCL_RATES: Record<number, number> = Object.fromEntries(
  Array.from({ length: ((SEA_MAX_TABLE_WEIGHT_KG - SEA_MIN_CHARGEABLE_WEIGHT_KG) / SEA_WEIGHT_STEP_KG) + 1 }, (_, index) => {
    const weight = Number((SEA_MIN_CHARGEABLE_WEIGHT_KG + (index * SEA_WEIGHT_STEP_KG)).toFixed(1));
    const price = Number((200 + (index * 4.5)).toFixed(2));
    return [weight, price];
  })
);

export function lookupShippingRate(originRegion: OriginRegion, weightKg: number): ShippingRateQuote {
  const rateBand: ShippingRateBand = weightKg <= 2 ? "document" : "parcel";
  const roundedWeight = roundUpToHalf(weightKg);
  const table = rateBand === "document"
    ? originRegion === "east"
      ? EAST_MALAYSIA_DOCUMENT_RATES
      : WEST_MALAYSIA_DOCUMENT_RATES
    : originRegion === "east"
      ? EAST_MALAYSIA_PARCEL_RATES
      : WEST_MALAYSIA_PARCEL_RATES;

  const availableWeights = Object.keys(table).map(Number).sort((left, right) => left - right);
  const rateWeightKg = Math.min(Math.max(roundedWeight, availableWeights[0]), availableWeights[availableWeights.length - 1]);
  const shippingFee = table[rateWeightKg] ?? table[availableWeights[availableWeights.length - 1]];

  return {
    originRegion,
    rateBand,
    rateWeightKg,
    shippingFee: Number(shippingFee.toFixed(2)),
  };
}

export function lookupSeaShippingRate(originRegion: OriginRegion, weightKg: number): ShippingRateQuote {
  const roundedWeight = roundUpToHalf(weightKg);
  const rateWeightKg = Math.min(Math.max(roundedWeight, SEA_MIN_CHARGEABLE_WEIGHT_KG), SEA_MAX_TABLE_WEIGHT_KG);
  const table = originRegion === "east" ? SEA_EAST_MALAYSIA_LCL_RATES : SEA_WEST_MALAYSIA_LCL_RATES;
  const shippingFee = table[rateWeightKg] ?? table[SEA_MAX_TABLE_WEIGHT_KG];

  return {
    originRegion,
    rateBand: "sea_lcl",
    rateWeightKg,
    shippingFee: Number(shippingFee.toFixed(2)),
  };
}

export function formatShippingRateLabel(quote: ShippingRateQuote): string {
  if (quote.rateBand === "sea_lcl") {
    const regionLabel = quote.originRegion === "east" ? "East Malaysia" : "West Malaysia";
    return `${regionLabel} Sea LCL rate at ${quote.rateWeightKg.toFixed(1)} kg`;
  }
  const regionLabel = quote.originRegion === "east" ? "East Malaysia" : "West Malaysia";
  const bandLabel = quote.rateBand === "document" ? "Document" : "Parcel";
  return `${regionLabel} ${bandLabel} rate at ${quote.rateWeightKg.toFixed(1)} kg`;
}

function roundUpToHalf(weightKg: number): number {
  return Math.ceil(weightKg * 2) / 2;
}